cuqi/experimental/mcmc/_hmc.py

```diff
@@ -0,0 +1,401 @@
+        The optimal acceptance rate to reach if using adaptive step size.
+        Suggested values are 0.6 (default) or 0.8 (as in stan).
+
+    callback : callable, *Optional*
+        If set this function will be called after every sample.
+        The signature of the callback function is
+        `callback(sample, sample_index)`,
+        where `sample` is the current sample and `sample_index` is the index of
+        the sample.
+        An example is shown in demos/demo31_callback.py.
+
+    Example
+    -------
+    .. code-block:: python
+
+        # Import cuqi
+        import cuqi
+
+        # Define a target distribution
+        tp = cuqi.testproblem.WangCubic()
+        target = tp.posterior
+
+        # Set up sampler
+        sampler = cuqi.experimental.mcmc.NUTSNew(target)
+
+        # Sample
+        sampler.warmup(5000)
+        sampler.sample(10000)
+
+        # Get samples
+        samples = sampler.get_samples()
+
+        # Plot samples
+        samples.plot_pair()
+
+    After running the NUTS sampler, run diagnostics can be accessed via the 
+    following attributes:
+
+    .. code-block:: python
+
+        # Number of tree nodes created each NUTS iteration
+        sampler.num_tree_node_list
+
+        # Step size used in each NUTS iteration
+        sampler.epsilon_list
+
+        # Suggested step size during adaptation (the value of this step size is
+        # only used after adaptation). The suggested step size is None if 
+        # adaptation is not requested.
+        sampler.epsilon_bar_list
+
+    """
+    def __init__(self, target, initial_point=None, max_depth=15,
+                 adapt_step_size=True, opt_acc_rate=0.6, **kwargs):
+        super().__init__(target, initial_point=initial_point, **kwargs)
+
+        # Assign parameters as attributes
+        self.max_depth = max_depth
+        self.adapt_step_size = adapt_step_size
+        self.opt_acc_rate = opt_acc_rate
+        
+        # Set current point 
+        self.current_point = self.initial_point
+
+        # Initialize epsilon and epsilon_bar
+        # epsilon is the step size used in the current iteration
+        # after warm up and one sampling step, epsilon is updated
+        # to epsilon_bar for the remaining sampling steps.
+        self._epsilon = None
+        self._epsilon_bar = None
+
+        # Arrays to store acceptance rate
+        self._acc = [None]
+
+        # NUTS run diagnostic:
+        # number of tree nodes created each NUTS iteration
+        self._num_tree_node = 0
+        # Create lists to store NUTS run diagnostics
+        self._create_run_diagnostic_attributes()
+
+    #=========================================================================
+    #================== Implement methods required by SamplerNew =============
+    #=========================================================================
+    def validate_target(self):
+        pass #TODO: target needs to have logpdf and gradient methods
```

Comment:
I would suggest adding a try catch here, simply attemping to evaluating _nuts_target and if the output is valid, accept the target.

Reply:
Thank you for the suggestion, I added the try-catch phrase.